### DISCUSSION TITLE
Add sha256 when updating docker images

### DISCUSSION THREAD

**Discussion by vlagorce** [renovatebot/renovate#16081] [Request Help] [answered]
### What would you like Renovate to be able to do?

1 - It would be great (mainly for security purpose) if renovate can add the sha256 of the updated image even it was not present.

For instance in a Dockerfile
```Dockerfile
FROM my-repo/my-image:1.0
```
```Dockerfile
FROM my-repo/my-image:2.0@sha256:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
2 - On the other hand would it be possible to add the "longest" more precise version if not define. When you have an update having only the sha256, it's hard to valdiate the version updated.
```Dockerfile
FROM my-repo/my-image@sha256:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
```Dockerfile
FROM my-repo/my-image:2.0@sha256:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

When using a tools like renovate, and having good practices on the docker images tagging there is no reason to use latest nor any short version. Using the "complete" version + the sha is more secure and easier to review.

### If you have any ideas on how this should be implemented, please tell us here.

For 1- I assume the manager have everything. 

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by HonkingGoose** [comment, marked as answer]
I think you're looking for this:

```json
{
  "docker": {
    "pinDigests": true
  }
}
```

Or you can put `docker:pinDigests` in your `extends` array.

**References:**

- [Renovate docs, Docker digest pinning](https://docs.renovatebot.com/docker/#digest-pinning)
- [Renovate docs, `docker:pinDigests`](https://docs.renovatebot.com/presets-docker/#dockerpindigests)

**Comment by vlagorce** [comment]
Sorry for the noise. Indeed it is what I was looking for.
Any idea if the version can be added ?. I did not see any preset/config regarding docker version except blocking major update.

**Reply by rarkins** [reply]
If the version is already present then it will be retained. But if it then it won't be added (and it's not an easy task for reasons related to the Docker API)

**Comment by HonkingGoose** [comment]
> When using a tools like renovate, and having good practices on the docker images tagging there is no reason to use latest nor any short version. Using the "complete" version + the sha is more secure and easier to review.

@rarkins do we want to update our `config:base` in a `breaking` change, so that Renovate will always use the complete version + SHA by default? Or maybe we can use a `config:recommended` preset which includes security best practices like this?

**Reply by viceice** [reply]
i think renovate can't pin the version, only the digest

**Reply by vlagorce** [reply]
Security wise I will tend to force to pin the digest (config:base) But I'm sure some user have different practices in private registry.
A config:recommended should be enough.

**Reply by rarkins** [reply]
I agree with you that everyone should do this, however it's too opinionated even for our config:base (lots of users will be annoyed, just like if we pin versions by default)

**Reply by vlagorce** [reply]
I'm wondering how the preset [docker:disableMajor](https://docs.renovatebot.com/presets-docker/#dockerdisablemajor) works. (I will dig in the code later on).
Renovate will have to resolve the `currentVersion` and the `newVersion`. Or this is only working when the version is explicitly set.

**Reply by viceice** [reply]
it's only working, when the version is alredy set
